src/pyclass/guard.rs

```diff
@@ -0,0 +1,745 @@
+/// ```
+///
+/// See also [`PyClassGuardMut`] and the [guide] for more information.
+///
+/// [Interior Mutability]:
+///     https://doc.rust-lang.org/book/ch15-05-interior-mutability.html
+///     "RefCell<T> and the Interior Mutability Pattern - The Rust Programming
+///     Language"
+/// [guide]: https://pyo3.rs/latest/class.html#bound-and-interior-mutability
+///     "Bound and interior mutability"
+#[repr(transparent)]
+pub struct PyClassGuard<'a, T: PyClass> {
+    ptr: NonNull<ffi::PyObject>,
+    marker: PhantomData<&'a Py<T>>,
+}
+
+impl<'a, T: PyClass> PyClassGuard<'a, T> {
+    pub(crate) fn try_borrow(obj: &'a Py<T>) -> Result<Self, PyBorrowError> {
+        Self::try_from_class_object(obj.get_class_object())
+    }
+
+    fn try_from_class_object(obj: &'a PyClassObject<T>) -> Result<Self, PyBorrowError> {
+        obj.ensure_threadsafe();
+        obj.borrow_checker().try_borrow().map(|_| Self {
+            ptr: NonNull::from(obj).cast(),
+            marker: PhantomData,
+        })
+    }
+
+    pub(crate) fn as_class_object(&self) -> &'a PyClassObject<T> {
+        // SAFETY: `ptr` by construction points to a `PyClassObject<T>` and is
+        // valid for at least 'a
+        unsafe { self.ptr.cast().as_ref() }
+    }
+}
+
+impl<'a, T, U> PyClassGuard<'a, T>
+where
+    T: PyClass<BaseType = U>,
+    U: PyClass,
+{
+    /// Borrows a shared reference to `PyClassGuard<T::BaseType>`.
+    ///
+    /// With the help of this method, you can access attributes and call methods
+    /// on the superclass without consuming the `PyClassGuard<T>`. This method
+    /// can also be chained to access the super-superclass (and so on).
+    ///
+    /// # Examples
+    /// ```
+    /// # use pyo3::prelude::*;
+    /// #[pyclass(subclass)]
+    /// struct Base {
+    ///     base_name: &'static str,
+    /// }
+    /// #[pymethods]
+    /// impl Base {
+    ///     fn base_name_len(&self) -> usize {
+    ///         self.base_name.len()
+    ///     }
+    /// }
+    ///
+    /// #[pyclass(extends=Base)]
+    /// struct Sub {
+    ///     sub_name: &'static str,
+    /// }
+    ///
+    /// #[pymethods]
+    /// impl Sub {
+    ///     #[new]
+    ///     fn new() -> (Self, Base) {
+    ///         (Self { sub_name: "sub_name" }, Base { base_name: "base_name" })
+    ///     }
+    ///     fn sub_name_len(&self) -> usize {
+    ///         self.sub_name.len()
+    ///     }
+    ///     fn format_name_lengths(slf: PyClassGuard<'_, Self>) -> String {
+    ///         format!("{} {}", slf.as_super().base_name_len(), slf.sub_name_len())
+    ///     }
+    /// }
+    /// # Python::attach(|py| {
+    /// #     let sub = Py::new(py, Sub::new()).unwrap();
+    /// #     pyo3::py_run!(py, sub, "assert sub.format_name_lengths() == '9 8'")
+    /// # });
+    /// ```
+    pub fn as_super(&self) -> &PyClassGuard<'a, U> {
+        // SAFETY: `PyClassGuard<T>` and `PyClassGuard<U>` have the same layout
+        unsafe { &*ptr_from_ref(self).cast() }

```

Comment:
I wonder if we should add a `non_null_from_ref` to match the soon-to-be-stabilized `NonNull::from_ref` helper (or we could have a private extension trait to add `from_ref` to `NonNull` on versions before 1.89?)


```suggestion
        unsafe { non_null_from_ref(self).cast().as_ref() }
```